url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/file/FORMFile.java

```diff
@@ -0,0 +1,203 @@
+    		}
+    		this.chunks.add(cChunk);
+    		pointer = cChunk.getEndPosition();
+    	}
+    }
+
+	public void deleteChunk(Chunk chunk, RandomAccessFile raf) throws IOException {
+		if (!this.getChunks().contains(chunk)) {
+			throw new IOException(KlangConstants.ERROR_NO_SUCH_CHUNK);
+		}
+		if ( (chunk.getStartPosition() == 0) && (chunk.getEndPosition() >= this.getFile().length()) ) {
+			throw new IOException(KlangConstants.ERROR_CHUNK_IS_FILE);
+		}
+		long end = KlangUtil.adjustForPadByte(chunk.getEndPosition());
+		KlangUtil.deleteFromFile(this.getFile(), chunk.getStartPosition(), end);
+		reparseFile(raf);
+	}
+	
+	public void addChunkFromFile(File theFile, String name, int position, RandomAccessFile raf) throws IOException {
+		//only FORM chunks can be added at root level
+		if (!name.equals("FORM")) {
+			throw new IOException(KlangConstants.ERROR_NOT_FORM);
+		}
+		if (position < 1) {
+			position = 1;
+		}
+		if (position > this.getChunks().size() + 1) {
+			position = this.getChunks().size() + 1;
+		}
+		long start = -1;
+		if (position == 1) {
+			start = 0;
+		} else {
+			start = this.getChunks().get(position - 2).getEndPosition();
+		}
+		//first, add header to a tempfile, then copy original file and possibly a pad byte
+		File tempFile = File.createTempFile("klang", ".data");
+		RandomAccessFile tempraf = new RandomAccessFile(tempFile, "rw");
+		try {
+			tempraf.seek(0);
+			//header
+			FourCC fccname = new FourCC();
+			fccname.setValueFromString(name);
+			fccname.writeValueToFile(tempraf);
+			Int4ByteUnsignedBE size = new Int4ByteUnsignedBE();
+			Long sizeval = KlangUtil.adjustForPadByte(theFile.length());
+			size.setValue(sizeval);
+			size.writeValueToFile(tempraf);
+		} catch (BadValueException err) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		//now write file to temp file
+		KlangUtil.appendToFile(theFile, tempFile);
+		if (KlangUtil.adjustForPadByte(theFile.length()) != theFile.length()) {
+			//add pad byte
+			KlangUtil.appendToFile(new byte[1], tempFile);
+		}
+		//now merge tempfile into file
+		KlangUtil.insertIntoFile(tempFile, this.getFile(), start);
+		tempFile.delete();
+		reparseFile(raf);
+	}
+	
+	public void addChunkFromArray(byte[] bytes, String name, int position, RandomAccessFile raf) throws IOException {
+		//only FORM chunks can be added at root level
+		if (!name.equals("FORM")) {
+			throw new IOException(KlangConstants.ERROR_NOT_FORM);
+		}
+		if (position < 1) {
+			position = 1;
+		}
+		if (position > this.getChunks().size() + 1) {
+			position = this.getChunks().size() + 1;
+		}
+		long start = -1;
+		if (position == 1) {
+			start = 0;
+		} else {
+			start = this.getChunks().get(position - 2).getEndPosition();
+		}
+		//first, add header to a tempfile, then add bytes and possibly a pad byte
+		File tempFile = File.createTempFile("klang", ".data");
+		RandomAccessFile tempraf = new RandomAccessFile(tempFile, "rw");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *tempraf*. Currently, there are execution paths that do not contain closure statements, for example, when FourCC.writeValueToFile throws an exception. Ensure *tempraf* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).